[Merged by Bors] - feat(Quiver/Path/Vertices): add splitting lemmas and decomposition results

or4nge19 opened the pull request:
feat(Quiver/Path/Vertices): add splitting lemmas and decomposition results
---
introduce splitting and decomposition results (exists_decomp_at_length, exists_decomp_of_mem_vertices, split_at_vertex needed to reason about the list of vertices of a path and to split paths at a chosen vertex. Part of Perron-Frobenius theorem pre-requisites
see https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/Formalizing.20Perron-Frobenius/with/525516636
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

or4nge19 committed:
feat(combinatorics): Define vertices of a Quiver.Path

This PR is linked to #27315 and #27325. It's part of PRing the formalization of core PF theorem here https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/Formalizing.20Perron-Frobenius/with/525516636

This adds:

Quiver.Path.vertices: A List V representing the sequence of vertices along a path.
Quiver.Path.vertices_set: The Set of vertices visited.
Quiver.Path.vertices_finset: The Finset of vertices visited.
Quiver.Path.init_finset: The Finset of vertices excluding the final vertex of the path, needed for defining simple paths and cycles.

vertices_comp: A proof that the vertices of a composed path p.comp q are p.vertices.dropLast ++ q.vertices.
vertices_length: The length of the vertices list is p.length + 1.
Lemmas connecting vertices to standard list operations like head?, getLast, and get.
Lemmas for membership: start_mem_vertices and end_mem_vertices.

This is the prerequisite for defining path simplicity and formalizing cycle extraction, as used in the formalization of the Perron-Frobenius theorem

or4nge19 committed:
Update Vertices.lean

removed unused import

or4nge19 committed:
Update Vertices.lean

added end_map

or4nge19 committed:
Revert "Update Vertices.lean"

This reverts commit a706e88a95d037a65856bf20201532cc126b71f9.

or4nge19 committed:
Merge branch 'master' into mc_pf_quiver_vertices

or4nge19 committed:
Update Mathlib/Combinatorics/Quiver/Path/Vertices.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

or4nge19 committed:
Update Vertices.lean

turned the verticesSet, verticesFinset and initFinset into [simp] lemmas

or4nge19 committed:
fix lint

or4nge19 committed:
Merge branch 'master' into mc_pf_quiver_vertices

or4nge19 committed:
Merge branch 'leanprover-community:master' into mc_pf_quiver_vertices

ocfnash committed:
Merge branch 'master' into mc_pf_quiver_vertices

ocfnash committed:
Fix `Mathlib.lean`

Achieved by running:
`lake exe mk_all`

or4nge19 committed:
Merge branch 'leanprover-community:master' into mc_pf_quiver_vertices

or4nge19 committed:
address comments

ok to all :) thanks so much

or4nge19 committed:
minor fix

remove unused string

or4nge19 committed:
Merge branch 'leanprover-community:master' into mc_pf_quiver_vertices

or4nge19 committed:
implements all comments, move length_toPath to quiver.Path

or4nge19 committed:
fix unused import

or4nge19 committed:
fix deprecated imports

or4nge19 committed:
feat(Quiver/Path/Vertices): add splitting lemmas and decomposition results

introduce splitting and decomposition results (exists_decomp_at_length, exists_decomp_of_mem_vertices, split_at_vertex needed to reason about the list of vertices of a path and to split paths at a chosen vertex. Part of Perron-Frobenius theorem pre-requisites
see

or4nge19 committed:
remove #lint

or4nge19 committed:
Merge branch 'master' into mc_pf_quiver_vertices

or4nge19 committed:
cosmetics

or4nge19 committed:
Merge branch 'mc_pf_quiver_vertices' of https://github.com/or4nge19/mathlib4 into mc_pf_quiver_vertices

or4nge19 committed:
add lemmas and make names more idiomatic

or4nge19 committed:
Merge branch 'leanprover-community:master' into mc_pf_quiver_vertices

ocfnash commented on Mathlib/Data/List/Basic.lean:
```diff
@@ -1287,4 +1288,22 @@ lemma left_le_of_mem_range' {a b s x : ℕ} (hx : x ∈ List.range' a b s) : a 
+section split
+
+/-- Any `x ∈ l` gives a decomposition `l = l₁ ++ x :: l₂`. -/
+lemma exists_mem_split {l : List α} {x : α} (h : x ∈ l) :
```
You can drop this lemma: as I discovered by using `exact?`, it already exists as (one direction of) [List.mem_iff_append](https://leanprover-community.github.io/mathlib4_docs/Init/Data/List/Lemmas.html#List.mem_iff_append).

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -5,7 +5,7 @@ Authors: Matteo Cipollina
 
 import Mathlib.Algebra.Order.Group.Nat
-import Mathlib.Combinatorics.Quiver.Path
+import Mathlib.Combinatorics.Quiver.Path--.Decomposition
```
```suggestion
import Mathlib.Combinatorics.Quiver.Path
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -160,7 +160,325 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
   simpa [h₁] using h₂
 
+/-!  ### Path vertices decomposition -/
+
+section
+
+variable {V : Type*} [Quiver V]
+
+open List
```
You already have these `variable`s, and `List` is already open. OTOH I think (almost?) every lemma below consumes a path between two vertices so how about:
```suggestion
variable {a b : V} (p : Path a b)
```
and then you can drop these arguments from all the lemma statements.

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -160,7 +160,325 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+
+/-- Given a path `p : Path a b` and an index `n ≤ p.length`,
+    we can split `p = p₁.comp p₂` with `p₁.length = n`. -/
+theorem exists_comp_of_length_le {a b : V} (p : Path a b) {n : ℕ} (hn : n ≤ p.length) :
```
```suggestion
theorem exists_eq_comp_of_le_length {a b : V} (p : Path a b) {n : ℕ} (hn : n ≤ p.length) :
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -160,7 +160,325 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+      p = p₁.comp p₂ ∧ p₁.length = n := by
+  induction p generalizing n with
+  | nil =>
+      have h : n = 0 := by simp_all only [length_nil, nonpos_iff_eq_zero]
+      subst h
+      exact ⟨a, Path.nil, Path.nil, by simp only [comp_nil], rfl⟩
+  | cons p' e ih =>
+      rename_i c
+      rw [length_cons] at hn
+      rcases (Nat.le_succ_iff).1 hn with h | h
+      · rcases ih h with ⟨d, p₁, p₂, hp, hl⟩
+        refine ⟨d, p₁, p₂.cons e, ?_, hl⟩
+        simp; rw [hp]
+      · subst h
+        refine ⟨c, p'.cons e, Path.nil, ?_, ?_⟩
+        all_goals simp
```
What you had is fine, but I claim the following is a mild improvement style-wise:
```suggestion
    obtain ⟨rfl⟩ : n = 0 := by simpa using hn
    exact ⟨a, Path.nil, Path.nil, by simp, rfl⟩
  | @cons _ c p' e ih =>
    rw [length_cons] at hn
    rcases (Nat.le_succ_iff).1 hn with h | rfl
    · obtain ⟨d, p₁, p₂, hp, hl⟩ := ih h
      exact ⟨d, p₁, p₂.cons e, by simp [hp], hl⟩
    · exact ⟨c, p'.cons e, Path.nil, by simp, by simp⟩
```

or4nge19 replied:
thanks!

ocfnash reviewed (commented):
Just some quick remarks for now; I'll return tomorrow.

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -160,7 +160,325 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+  simp [head?_cons, Option.some.injEq] at this
+  exact this
+
+lemma vertices_head_eq_start {a b : V} (p : Path a b) :
+    p.vertices.head (vertices_ne_nil p) = a := by
+  induction p with
+  | nil => simp only [vertices_nil, head_cons]
+  | cons p' _ ih =>
+    simp [vertices_cons, concat_eq_append]
+    have : p'.vertices ≠ [] := vertices_ne_nil p'
+    simp [head_append_of_ne_nil this]
```
This is just `vertices_head_eq` higher up.
```suggestion
```
(in fact `vertices_head_eq` is slightly more extensible)

or4nge19 commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -160,7 +160,325 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+  refine ⟨v, p₁, p₂, rfl, hlen, ?_⟩
+  simp [hlen]
+
+lemma not_mem_vertices_dropLast_of_split_at_vertex
+    [DecidableEq V] {a b v : V} {p : Path a b}
+    (h_v_in_p : v ∈ p.vertices) :
+    let i  := p.vertices.idxOf v
+    let hi := List.idxOf_lt_length_iff.2 h_v_in_p
+    ∀ (c : V) (p_prefix : Path a c) (p_suffix : Path c b)
+      (_ : p = p_prefix.comp p_suffix)
+      (_ : p_prefix.length = i)
+      (_ : c = p.vertices.get ⟨i, hi⟩),
+      v ∉ p_prefix.vertices.dropLast := by
```
I fully agree, let's remove these two lemmas. the results in Quiver that would be hard to replace in this PF pipeline are the theorems, I don't remember these two are much used, except in the next Quiver PR on the API for IsSimple predicate for paths, so we we'll see soon.

or4nge19 committed:
address comments

or4nge19 committed:
fix comments

remove List lemma and use the correct one in exists_comp_of_mem_vertices

or4nge19 committed:
fix lint

or4nge19 committed:
Remove unused split section in List module

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -160,7 +160,201 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+
+/-- If a vertex `v` occurs in the list of vertices of a path `p : Path a b`, then `p` can be
+decomposed as a concatenation of a subpath from `a` to `v` and a subpath from `v` to `b`. -/
+theorem exists_comp_of_mem_vertices {a b v : V} (p : Path a b)
+  (h : v ∈ p.vertices) : ∃ (p₁ : Path a v) (p₂ : Path v b), p = p₁.comp p₂ := by
```
Let's use the `variable`s defined above to shorten this:
```suggestion
theorem exists_comp_of_mem_vertices {v : V} (h : v ∈ p.vertices) :
    ∃ (p₁ : Path a v) (p₂ : Path v b), p = p₁.comp p₂ := by
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -160,7 +160,201 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+
+/-- `split_at_vertex` decomposes a path `p` at the vertex sitting in
+    position `i` of its `vertices` -/
+theorem split_at_vertices_index (i : ℕ)
+    (hi : i < p.vertices.length) :
+    ∃ (v : V) (p₁ : Path a v) (p₂ : Path v b),
```
```suggestion
theorem exists_eq_comp_and_length_eq_of_lt_length (i : ℕ) (hi : i < p.vertices.length) :
    ∃ (v : V) (p₁ : Path a v) (p₂ : Path v b),
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -160,7 +160,201 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+    exact Nat.le_of_lt_succ hi
+  obtain ⟨v, p₁, p₂, rfl, rfl⟩ := exists_eq_comp_of_le_length p hi_le_len
+  exact ⟨v, p₁, p₂, rfl, rfl, by simp⟩
+
+/-- Given vertices lists from a path composition,
+the prefix path’s vertices is a prefix of the full path’s vertices. -/
+lemma vertices_dropLast_isPrefix_of_vertices_eq
+    {V : Type*} [Quiver V]
+    {a b c : V} {p : Path a b} {p₁ : Path a c} {p₂ : Path c b}
+    (h : p.vertices = p₁.vertices.dropLast ++ p₂.vertices) :
+    p₁.vertices.dropLast.IsPrefix p.vertices := by
+  rw [h]; exact prefix_append p₁.vertices.dropLast p₂.vertices
```
This is just a wrapper around `List.prefix_append` so we should drop it:
```suggestion
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -160,7 +160,201 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+    p₁.vertices.dropLast.IsPrefix p.vertices := by
+  rw [h]; exact prefix_append p₁.vertices.dropLast p₂.vertices
+
+lemma mem_vertices_eq_end_or_mem_dropLast {a b c : V} (p : Path a b) (h : c ∈ p.vertices) :
+  c = b ∨ c ∈ p.vertices.dropLast := by
+  induction p with
+  | nil =>
+    simp [vertices_nil] at h
+    subst h
+    simp
+  | cons p' e ih =>
+    rename_i b'
+    have h' : c ∈ p'.vertices ∨ c = b' := by
+      simpa using (mem_vertices_cons p' e).1 h
+    cases h' with
+    | inl h_in_p' =>
+      have ih' := ih h_in_p'
+      cases ih' with
+      | inl h_eq_p'_end =>
+        right
+        subst h_eq_p'_end
+        simp
+      | inr h_in_p'_dropLast =>
+        simp_all
+    | inr h_eq_b =>
+      left
+      exact h_eq_b
```
This is really a lemma about lists. There is just one detail which is special to the situation of paths in a quiver which is:
```lean
@[simp] lemma dropLast_append_end_eq :
    p.vertices.dropLast ++ [b] = p.vertices := by
  simp_rw [← p.vertices_getLast p.vertices_ne_nil, dropLast_concat_getLast]
```
Given this we can prove `mem_vertices_eq_end_or_mem_dropLast` quite efficiently as follows:
```lean
example {c : V} (h : c ∈ p.vertices) :
    c ∈ p.vertices.dropLast ∨ c = b := by
  rw [← dropLast_append_end_eq] at h
  simpa [-dropLast_append_end_eq] using h
```
So I suggest adding `dropLast_append_end_eq` (somewhere higher up in the file) and then dropping `mem_vertices_eq_end_or_mem_dropLast` altogether.

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -160,7 +160,201 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+      left
+      exact h_eq_b
+
+/-- If we have a path p from a to b with c ∈ p.vertices,
+    and c is not the end vertex b, then it appears in a proper prefix of the path. -/
+lemma exists_prefix_of_mem_vertices_ne_end [DecidableEq V] {a b c : V}
+    (p : Path a b) (h : c ∈ p.vertices) (h_ne : c ≠ b) :
+  ∃ (p₁ : Path a c) (p₂ : Path c b), p = p₁.comp p₂ := by
+  have h_cases := mem_vertices_eq_end_or_mem_dropLast p h
+  cases h_cases with
+  | inl h_eq =>
+      contradiction
+  | inr h_mem_tail =>
+      let i := p.vertices.idxOf c
+      have hi : i < p.vertices.length := List.idxOf_lt_length_iff.2 h
+      obtain ⟨v, p₁, p₂, h_comp, h_len, h_c_eq⟩ := split_at_vertices_index p i hi
+      have hvc : v = c := by
+        rw [h_c_eq]
+        simp [i]
+      subst hvc
+      exact ⟨p₁, p₂, h_comp⟩
+
```
Let's just drop this lemma: the hypothesis `h_ne` is redundant (to witness this, replace the `contradiction` tactic by just `subst h_eq; exact ⟨p, Path.nil, by simp⟩`) and it is just a weaker form of the lemma immediately below.
```suggestion
```

ocfnash commented on Mathlib/Combinatorics/Quiver/Path/Vertices.lean:
```diff
@@ -160,7 +160,201 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+      exact ⟨p₁, p₂, h_comp⟩
+
+/-- Split a path at the *last* occurrence of a vertex. -/
+theorem exists_decomp_of_mem_vertices_not_mem_tail
+    [DecidableEq V] {a b x : V} (p : Path a b) (hx : x ∈ p.vertices) :
+    ∃ (p₁ : Path a x) (p₂ : Path x b),
+      p = p₁.comp p₂ ∧ x ∉ p₂.vertices.tail := by
```
```suggestion
theorem exists_eq_comp_and_notMem_tail_of_mem_vertices {x : V} (hx : x ∈ p.vertices) :
    ∃ (p₁ : Path a x) (p₂ : Path x b),
      p = p₁.comp p₂ ∧ x ∉ p₂.vertices.tail := by
```

ocfnash reviewed (commented):
Thanks, I think very close after these latest suggestions!

or4nge19 committed:
Merge branch 'leanprover-community:master' into mc_pf_quiver_vertices

or4nge19 committed:
adopt review's suggestions

ocfnash committed:
Final tweaks (lemma placement, style nits, golf)

The lemma `Quiver.Path.exists_eq_comp_and_notMem_tail_of_mem_vertices` looks
like it could be broken down into API for `Quiver.Path.vertices` and for `List`
but the statement is good and the proof is maintainable so we can live with it
as it stands.

ocfnash commented:
-awaiting-author

ocfnash commented:
bors merge

ocfnash commented:
> thanks again for your time and all patience you had :)

You're very welcome!
